File: social-outing-planner/server/routes/planning.py
from flask import Blueprint, request, jsonify
from datetime import datetime
from server.config import Config
from server.storage import JsonStorage
from server.models import Event
from server.utils import generate_id, token_required

planning_bp = Blueprint('planning', __name__)

@planning_bp.route('/events', methods=['GET'])
@token_required
def get_events(current_user):
    try:
        events = JsonStorage.read(Config.EVENTS_FILE)
        # Return events where user is the creator OR is in the sharedWith list
        user_events = [
            event for event in events 
            if event.get('userId') == current_user['userId'] or 
            current_user['userId'] in event.get('sharedWith', [])
        ]
        return jsonify(user_events)
    except Exception as error:
        return jsonify({'error': str(error)}), 500

@planning_bp.route('/events', methods=['POST'])
@token_required
def create_event(current_user):
    try:
        data = request.get_json()
        events = JsonStorage.read(Config.EVENTS_FILE)
        
        # Extract sharedWith from data, default to empty list if not provided
        # shared_with = data.get('sharedWith', []) 
        # Model handles defaults if not provided, but we pass data directly.
        
        event_data = {
            '_id': generate_id(),
            **data,
            'userId': current_user['userId'],
            'createdBy': current_user['username']
        }
        event = Event.from_dict(event_data)
        
        events.append(event.to_dict())
        JsonStorage.write(Config.EVENTS_FILE, events)
        
        return jsonify(event.to_dict())
    except Exception as error:
        return jsonify({'error': str(error)}), 400

@planning_bp.route('/events/<event_id>', methods=['DELETE'])
@token_required
def delete_event(current_user, event_id):
    try:
        events = JsonStorage.read(Config.EVENTS_FILE)
        
        # Find the event to check if user is the creator
        event_to_delete = next((e for e in events if e.get('_id') == event_id), None)
        
        if not event_to_delete:
            return jsonify({'error': 'Event not found'}), 404
        
        # Only allow the creator to delete the event
        if event_to_delete.get('userId') != current_user['userId']:
            return jsonify({'error': 'Only the event creator can delete this event'}), 403
        
        filtered_events = [event for event in events if event.get('_id') != event_id]
        JsonStorage.write(Config.EVENTS_FILE, filtered_events)
        
        return jsonify({'message': 'Event deleted successfully'})
    except Exception as error:
        return jsonify({'error': str(error)}), 500

File: social-outing-planner/rust-server/src/routes/planning.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::Deserialize;
use serde_json::Value;
use crate::{
    models::Event,
    state::AppState,
    utils::{generate_id, AuthUser},
};

// DTOs
#[derive(Deserialize)]
pub struct CreateEventRequest {
    #[serde(default)]
    title: String,
    #[serde(default)]
    date: String,
    #[serde(default, rename = "startTime")]
    start_time: String,
    #[serde(default, rename = "endTime")]
    end_time: String,
    #[serde(default)]
    description: String,
    #[serde(default, rename = "type")]
    event_type: String,
    location: Option<String>,
    duration: Option<serde_json::Value>,
    #[serde(default)]
    category: String,
    #[serde(default)]
    start: String,
    #[serde(default)]
    end: String,
    #[serde(default, rename = "allDay")]
    all_day: bool,
    resource: Option<Value>,
    #[serde(rename = "sharedWith")]
    shared_with: Option<Vec<String>>,
}

pub async fn get_events(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let events: Vec<Event> = state.events_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    
    let user_events: Vec<Event> = events.into_iter()
        .filter(|e| e.user_id == auth.user_id || e.shared_with.contains(&auth.user_id))
        .collect();
        
    Ok(Json(user_events))
}

pub async fn create_event(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(payload): Json<CreateEventRequest>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let mut events: Vec<Event> = state.events_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let new_event = Event {
        id: generate_id(),
        user_id: auth.user_id,
        created_by: auth.username,
        title: payload.title,
        date: payload.date,
        start_time: payload.start_time,
        end_time: payload.end_time,
        description: payload.description,
        event_type: payload.event_type,
        location: payload.location,
        duration: payload.duration,
        category: payload.category,
        start: payload.start,
        end: payload.end,
        all_day: payload.all_day,
        resource: payload.resource,
        shared_with: payload.shared_with.unwrap_or_default(),
        created_at: chrono::Utc::now().to_rfc3339(),
    };

    events.push(new_event.clone());
    state.events_store.write(&events).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(new_event))
}

pub async fn delete_event(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(event_id): Path<String>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let events: Vec<Event> = state.events_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    
    // Check ownership before deleting
    let event = events.iter().find(|e| e.id == event_id);
    if let Some(e) = event {
        if e.user_id != auth.user_id {
            return Err((StatusCode::FORBIDDEN, "Only the event creator can delete this event".to_string()));
        }
    } else {
        return Err((StatusCode::NOT_FOUND, "Event not found".to_string()));
    }

    let filtered_events: Vec<Event> = events.into_iter()
        .filter(|e| e.id != event_id)
        .collect();

    state.events_store.write(&filtered_events).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "message": "Event deleted successfully" })))
}
